spec/models/ability_spec.rb

```diff
@@ -89,11 +89,103 @@
+      context 'when user is not a speaker with event confirmed' do
+        let(:conference) { create(:conference) }
+
+        context 'when the registration is closed' do
+          let(:conference_with_closed_registration) { create(:conference) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference: conference_with_closed_registration)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference_with_closed_registration)) }
+        end
+
+        context 'when the registration period is not set' do
+          it{ should_not be_able_to(:new, Registration.new(conference: conference)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference)) }
+        end
+
+        context 'when user has already registered' do
+          let(:user_registration) { create(:registration, conference: conference, user: user) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference_id: user_registration.conference_id)) }
+          it{ should_not be_able_to(:create, Registration.new(conference_id: user_registration.conference_id)) }
+        end
+
+        context 'when user has not registered and registrations are open with no registration_limit_exceeded' do
+          let(:registration_period) { create(:registration_period) }
+          let(:conference_with_open_registration_and_limit_not_excedeed) { create(:conference, registration_limit: '1', registration_period: registration_period) }
+
+          it{ should be_able_to(:new, Registration.new(conference: conference_with_open_registration_and_limit_not_excedeed)) }
+          it{ should be_able_to(:create, Registration.new(conference: conference_with_open_registration_and_limit_not_excedeed)) }
+        end
+
+        context 'when user has not registered and registrations are open with registration_limit_exceeded' do
+          let(:registration_period) { create(:registration_period) }
+          let(:conference_with_open_registration_and_limit_excedeed) { create(:conference, registration_limit: '1', registration_period: registration_period) }
+          let(:registration_for_user_with_open_registration_and_limit_excedeed) { create(:registration, conference: conference_with_open_registration_and_limit_excedeed, user: user2) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference_id: registration_for_user_with_open_registration_and_limit_excedeed.conference_id)) }
+          it{ should_not be_able_to(:create, Registration.new(conference_id: registration_for_user_with_open_registration_and_limit_excedeed.conference_id)) }
+        end
+      end
+
+      context 'when user is a speaker with event confirmed' do
+        context 'when the registration period is not set' do
+          let(:conference_with_speaker_confirmed) { create(:conference) }
+          let(:event_with_speaker_confirmed) { create(:event, state: 'confirmed', program: conference_with_speaker_confirmed.program) }
```

Comment:
You are defining this in multiple places, it seems it should be defined once after line 136